### PR TITLE
Changing Distances adjoints to ChainRules syntax

### PR DESCRIPTION
**Pull request by theogf**
Using @adjoint disable the overloading of `rrule` see : https://github.com/JuliaDiff/ChainRulesCore.jl/issues/239#issuecomment-713560336
It makes it impossible to define new rules for different metrics.
I am aware that the long-term plan is to move out the rules to the respective packages but in the mean-time it would be nice for it to be solved (it blocks our development in https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/pull/208).

### CODE REVIEW

**Comment by DhairyaLGandhi** [comment]
Lgtm for the most part.

**Comment by DhairyaLGandhi** [comment]
`@adjoint` does take precedence over `rrule`, so you could also define it via the macro to change it later. Although I would ask what the use case for that would be?

**Comment by theogf** [comment]
Sorry, I made a mistake: trying JuliaGaussianProcesses/KernelFunctions.jl#208 with this branch does not seem to solve the problem.

Also precompiling this patch returns a surprising warning: 
```julia
┌ Warning: Error requiring `Distances` from `Zygote`
│   exception =
│    LoadError: error in method definition: function ChainRulesCore.rrule must be explicitly imported to be extended
...
```

**Comment by DhairyaLGandhi** [comment]
Yeah, so we usually move the adjoints over, but since ChainRules is a dep, you could add the import

**Comment by willtebbutt** [comment]
@theogf I know you've resolved this on your end now, but would you be up for pushing this PR through anyway? Seems like something that we should be doing throughout `Zygote` when the opportunity arises.

**Comment by devmotion** [comment]
It seems the adjoint definitions have to be updated to the ChainRules-syntax? In particular, I guess `nothing` should be replaced with the ChainRules-equivalents.

**Comment by theogf** [comment]
I officially don't know how to import rrule correctly via Requires...

**Comment by theogf** [comment]
I am not sure the failing checks are related to this PR...

**Comment by willtebbutt** [comment]
Doesn't look related to me. @theogf I reckon this is basically good to go after a patch bump. I'll bors it after that unless @DhairyaLGandhi has any objections?

**Review by DhairyaLGandhi — APPROVED** [review]

**Comment by DhairyaLGandhi** [comment]
There was a merge conflict, but lgtm

File: Project.toml
Comment (CarloLucibello) [inline Project.toml:3]:
```suggestion
version = "0.6.10"
```

File: src/lib/distances.jl
Comment (devmotion) [inline src/lib/distances.jl:68]: This definition calls `Zygote.pullback` to compute the pullback of a more stable modification of the forward pass. This seems unfortunate for everyone that would like to use the ChainRules definitions with another AD backend. But on the other hand, the definitions here are type piracy anyway and problematic in such cases where Zygote is loaded (possibly by some other dependency).

Ref: https://github.com/JuliaDiff/ChainRules.jl/issues/338, https://github.com/JuliaDiff/ChainRulesCore.jl/issues/68

Comment (ToucheSir) [reply src/lib/distances.jl:68]: @theogf up to revive this using the pattern mentioned in https://github.com/JuliaDiff/ChainRules.jl/issues/338#issuecomment-884772928?

Comment (theogf) [reply src/lib/distances.jl]: Done! Hopefully I understood correctly how it worked

File: src/lib/distances.jl
Comment (devmotion) [inline src/lib/distances.jl:2]: IIRC `NO_FIELDS` was deprecated a long time ago. Instead one should use just `NoTangent`.

File: src/lib/distances.jl
Comment (devmotion) [inline src/lib/distances.jl:4]: I think it's not a good idea to add these type piracies to Zygote. I assume it is better to move it to e.g. ChainRules or Distances (IIRC the main developers did not want to add rules...) right away. Otherwise I fear it will be not only surprising that Zygote adds these rules (and one has to load Zygote to get these chainrules definitions) but also more annoying if one moves it to another package later on (since one has to be careful to avoid method redefinitions etc.).

A temporary workaround could also be to define the rules in Zygote, but only for `ZygoteRuleConfig`.

Comment (theogf) [reply src/lib/distances.jl:4]: Sure, but the problem is that if I start to add these rules in ChainRules, will we be able to test them?

Comment (devmotion) [reply src/lib/distances.jl:4]: In ChainRules? There you would just test them with ChainRulesTestUtils. Actually, this should be done regardless of where they are defined to ensure that the rules are implemented properly, return the correct types, have the correct signature, are type stable etc.

Comment (mcabbott) [reply src/lib/distances.jl:4]: Xref https://github.com/JuliaArrays/FillArrays.jl/pull/153 maybe. I don't think CR.jl wants to provide rules for, and hence depend on, other packages. If they don't want to depend on CRC.jl, then the other option is to make tiny glue packages for the rules. Or to wait for Pkg to build in this ability. 

But setting up rules using ZygoteRuleConfig here can't hurt, and would make it easier to move them elsewhere later.

Comment (ToucheSir) [reply src/lib/distances.jl:4]: Would ChainRules be willing to take on Distances.jl as a dependency? I assumed the only path for an external package outside of upstreaming to Distances itself would be something like DistributionsAD.jl.

In the short-medium term interest of removing as many (ideally all) `adjoint`s from Zygote itself, using `ZygoteRuleConfig` is a great option.

Comment (devmotion) [reply src/lib/distances.jl:4]: Please not something like DistributionsAD :smile: It's a complete mess, full of type piracy, and to a large extent it contains workarounds and hacks that should have been added to the respective AD backends or other upstream packages such as Distributions or PDMats. The long term plan is to get rid of it as much as possible, and already all CR things were moved to Distributions. While it speeded up initial development to accumulate all the stuff in the package instead of fixing other packages, we now have to pay the price for it and spend a significant amount of time with cleaning it.

Comment (ToucheSir) [reply src/lib/distances.jl:4]: Not saying it's a good idea :P, more that it's the only option outside of hosting rules in the AD backend if upstream is not willing to accept PRs.

Comment (devmotion) [reply src/lib/distances.jl:4]: :stuck_out_tongue: 

IMO defining the rules for ZygoteRuleConfig only is probably the easiest and cleanest option if one wants to avoid (repeating) these discussions about whether it rather belongs to ChainRules/Distances/etc. here.

Comment (mzgubic) [reply src/lib/distances.jl:4]: >Would ChainRules be willing to take on Distances.jl as a dependency?

The policy is that ChainRules provides rules for functionality in Julia Base. Packages should depend on ChainRulesCore (which is super light and has no dependencies other than Julia Base) to define rules for their functionality. @oxinabox would have to approve exceptions, but I don't think I can remember any, apart from two in the early days that have since been moved out.

I agree that the use of ZygoteRuleConfig is the best solution for now. If we decide in the future where we want to put them it would be made very easy as well.

File: src/lib/distances.jl
Comment (devmotion) [inline src/lib/distances.jl:4]:
```suggestion
function rrule(::ZygoteRuleConfig, ::SqEuclidean, x::AbstractVector, y::AbstractVector)
```

File: src/lib/distances.jl
Comment (devmotion) [inline src/lib/distances.jl:13]:
```suggestion
function rrule(::ZygoteRuleConfig, ::typeof(colwise), s::SqEuclidean, x::AbstractMatrix, y::AbstractMatrix)
```

File: src/lib/distances.jl
Comment (devmotion) [inline src/lib/distances.jl:20]:
```suggestion
function rrule(::ZygoteRuleConfig, ::typeof(pairwise), s::SqEuclidean, x::AbstractMatrix, y::AbstractMatrix; dims::Int=2)
```

File: src/lib/distances.jl
Comment (devmotion) [inline src/lib/distances.jl:35]:
```suggestion
function rrule(::ZygoteRuleConfig, ::typeof(pairwise), s::SqEuclidean, x::AbstractMatrix; dims::Int=2)
```

File: src/lib/distances.jl
Comment (devmotion) [inline src/lib/distances.jl:50]:
```suggestion
function rrule(::ZygoteRuleConfig, ::Euclidean, x::AbstractVector, y::AbstractVector)
```

File: src/lib/distances.jl
Comment (devmotion) [inline src/lib/distances.jl:60]:
```suggestion
function rrule(::ZygoteRuleConfig, ::typeof(colwise), s::Euclidean, x::AbstractMatrix, y::AbstractMatrix)
```

**Review by devmotion — COMMENTED** [review]
I think it would also be good to add some tests with ChainRulesTestUtils.test_rrule.

**Comment by CarloLucibello** [comment]
needs a rebase, otherwise seems a sensible change

**Comment by devmotion** [comment]
I wonder if possibly maintainers of Distances who didn't want to add a dependency on ChainRulesCore might be fine with adding it as a weak dependency?

File: src/lib/distances.jl
Comment (mcabbott) [inline src/lib/distances.jl:79]: Maybe clearer without these lines, stop at `D, back = rrule_via_ad(config, ...)`? Or just `return rrule_via_ad(config, ...)`.

Comment (theogf) [reply src/lib/distances.jl:79]: Sure! I only wanted to check that this was the issue 😊

**Comment by ToucheSir** [comment]
CI is complaining about a `\Delta` not being defined, but it seems like it is?

**Pull request closed without merging**
